src/xdp/xsk.c

```diff
@@ -234,11 +234,34 @@ XskSignalReadyIo(
 
     KeAcquireSpinLock(&Xsk->Lock, &OldIrql);
     if ((Xsk->IoWaitFlags & ReadyFlags) != 0) {
+        Xsk->IoWaitFlags = 0;
```

Comment:
Updated to only remove the flag being completed.

Reply:
Actually, I _think_ this was a bug in the previous/existing code, because the waiting thread was clearing this too.

Reply:
You're right, this is (and was) safe because the waiting thread keeps track of its wait condition itself. The shared mask in the XSK is only needed to ensure the event is set once.

Reply:
I lied: you definitely cannot clear any of the bits here, since this field being nonzero prevents multiple threads entering the wait state simultaneously, which is a fundamental requirement of this wait implementation.

Reply:
Once you call `KeSetEvent` it's done. You've woken the waiting thread. At that point, you can safely have another thread come in, clear the event and wait.

Reply:
I agree with your sentiment, but that's not safe in our implementation. Easy example of why not:

Thread A sets up wait, doesn't actually call NtWaitForSingleObject (yet)
Thread B cancels wait, sets event
Thread C enters wait, clears event
Thread A calls NtWaitForSingleObject